### ISSUE TITLE
Table/Feature Request: various system_profiler virtual tables

### ISSUE THREAD

**Issue by FritzX6** [kolide/launcher#572]
## What is being requested?

The binary `/usr/sbin/system_profiler` can be used to gather various system data that is currently unavailable in the core osquery tables.

This issue proposes a host of new tables and modifications to existing tables to flesh out visibility gaps present in the core osquery project.

This binary was previously reverse engineered by @terracatta in constructing his iBridge Table: https://github.com/osquery/osquery/pull/5166

## What virtual tables are being proposed?
The following virtual tables apply only to macOS and can be populated using the `system_profiler NamedDataType` command:

- [ ] kolide_macos_system_info `SPHardwareDataType`
- [ ] kolide_macos_ibridge_configuration `SPiBridgeDataType`
- [ ] kolide_macos_configuration_profiles `SPConfigurationProfileDataType`

The following virtual tables are lower priority but should be simple once the framework is in place to shell out and parse the returned JSON of the above tables:

- [ ] kolide_macos_os_version `SPSoftwareDataType`
- [ ] kolide_macos_bluetooth_configuration `SPBluetoothDataType`
- [ ] kolide_macos_bluetooth_devices `SPBluetoothDataType`
- [ ] kolide_macos_ac_power_settings `SPPowerDataType`
- [ ] kolide_macos_gpu_info `SPDisplayDataType`
- [ ] kolide_macos_display_devices `SPDisplayDataType`
...

These outputs can accept various formatting options, including json and xml.

## Why is this valuable?

There are a number of data points which have been requested by customers which core osquery does not support, and that this pattern can retrieve. For Example:

- Activation Lock Status
- iBridge Information (in osquery but currently broken)
- Connected Bluetooth Devices

## What is the output of this binary?

```shell
➜  system_profiler SPHardwareDataType 
Hardware:

    Hardware Overview:

      Model Name: iMac Pro
      Model Identifier: iMacPro1,1
      Processor Name: 8-Core Intel Xeon W
      Processor Speed: 3.2 GHz
      Number of Processors: 1
      Total Number of Cores: 8
      L2 Cache (per Core): 1 MB
      L3 Cache: 11 MB
      Hyper-Threading Technology: Enabled
      Memory: 32 GB
      Boot ROM Version: 1037.60.58.0.0 (iBridge: 17.16.12551.0.0,0)
      Serial Number (system): C02VT3******
      Hardware UUID: E664D52B-5FBC-5A68-A0A5-******
      Activation Lock Status: Disabled
```

Passed with json formatting option:
```shell
➜  system_profiler SPHardwareDataType -json
{
  "SPHardwareDataType" : [
    {
      "_name" : "hardware_overview",
      "activation_lock_status" : "activation_lock_disabled",
      "boot_rom_version" : "1037.60.58.0.0 (iBridge: 17.16.12551.0.0,0)",
      "cpu_type" : "8-Core Intel Xeon W",
      "current_processor_speed" : "3.2 GHz",
      "l2_cache_core" : "1 MB",
      "l3_cache" : "11 MB",
      "machine_model" : "iMacPro1,1",
      "machine_name" : "iMac Pro",
      "number_processors" : 8,
      "packages" : 1,
      "physical_memory" : "32 GB",
      "platform_cpu_htt" : "htt_enabled",
      "platform_UUID" : "E664D52B-5FBC-5A68-******",
      "serial_number" : "C02VT3******"
    }
  ]
}%
```

system_profiler man output:
```shell
SYSTEM_PROFILER(8)        BSD System Manager's Manual       SYSTEM_PROFILER(8)

NAME
     system_profiler -- reports system hardware and software configuration.

SYNOPSIS
     system_profiler [-usage]
     system_profiler [-listDataTypes]
     system_profiler [-xml] dataType1 ... dataTypeN
     system_profiler [-xml] [-detailLevel level]
     system_profiler [-json] dataType1 ... dataTypeN
     system_profiler [-json] [-detailLevel level]

DESCRIPTION
     system_profiler reports on the hardware and software configuration of the system.  It can generate plain text reports or XML reports which can be opened with System Information.app, or JSON reports

     Progress and error messages are printed to stderr while actual report data is printed to stdout. Redirect stderr to /dev/null to suppress progress and error messages.

     The following options are available:

     -xml                Generates a report in XML format.  If the XML report is redirected to a file with a ".spx" suffix that file can be opened with System Information.app.

     -json               Generates a report in JSON format.

     -listDataTypes      Lists the available datatypes.

     -detailLevel level  Specifies the level of detail for the report:

                         mini          report with no personal information

                         basic         basic hardware and network information

                         full          all available information

     -timeout            Specifies the maximum time to wait in seconds for results.  If some information is not available within the specified time limit then an incomplete or partial report will be generated. The default timeout is 180 seconds. Specifying a
                         timeout of 0 means no timeout.

     -usage              Prints usage info and examples.

EXAMPLES
     system_profiler
       Generates a text report with the standard detail level.

     system_profiler -detailLevel mini
       Generates a short report containing no personal information.

     system_profiler -listDataTypes
       Shows a list of the available data types.

     system_profiler SPSoftwareDataType SPNetworkDataType
       Generates a text report containing only software and network data.

     system_profiler -xml > MyReport.spx
       Creates a XML file which can be opened by System Profiler.app

AUTHORS
     Apple Computer, Inc.

Darwin                           June 30, 2003                          Darwin
```

**Comment by directionless** [comment]
Hi @FritzX6 I started playing around with this. The underlying data feels very EAV oriented. The model seems to be that one calls `system_profiler` with any set of dataTypes to return.  For example: `system_profiler SPHardwareDataType  SPPowerDataType`.

I suspect it would be pretty reasonable to create an EAV table that supported something like `select ... where datatype in (...)`  This would also be pretty robust against data differences across OS versions.

But, I suspect hardcoding the columns and making a table per data type might be more useable.

Thoughts?

**Comment by groob** [comment]
IMO the proper place for these tables is osquery, because the iokit APIs provide access to the values, and it's much more performant to query that way. system_profiler is slow.

**Comment by directionless** [comment]
> the iokit APIs provide access to the values

I'm not super familiar with iokit. Are you sure all the values are there? Are there docs? Poking around, I mostly find `ioreg` (and the existing `iokit_registry` table). but that seems to not quite be the same. Given the naming, and man page snippets, maybe we get the objects from the register, then we query them.

**Comment by directionless** [comment]
Fixed in https://github.com/kolide/launcher/pull/574
